### DISCUSSION TITLE
textureSample in uniform control flow warning

### DISCUSSION THREAD

**Discussion by AshleyScirra** [gpuweb/gpuweb#2899] [Q&A] [answered]
Earlier on in WebGPU's development I built a multi-texturing fragment shader. Basically it samples from multiple 2D textures using a vec3 for the texture co-ordinate: X and Y for the sample position, and Z to select between a few possible texture bindings. The code is more or less like this, but extended to cover more textures:

```wgsl
var texXy : vec2<f32> = input.fragUV.xy;
var texIndex : f32 = input.fragUV.z;
var c : vec4<f32>;

if (texIndex < 0.5)
{
	c = textureSample(texture0, sampler0, texXy)
}
else
{
	c = textureSample(texture1, sampler1, texXy)
}

// ... output 'c' ...
```

This means we can bind multiple textures at once and let the fragment shader choose between them, avoiding the need for lots of setBindGroup() calls when we want to switch which texture is being sampled from.

However recently Chrome Canary started issuing a warning for this style of code:

> warning: 'textureSample' must only be called from uniform control flow

I note that this appears to only be a *warning* and shader compilation still succeeds and everything seems to render fine. However I'm wondering how much of a problem this is.

I know that the code is effectively uniform because for any given triangle `fragUV.z` is always set to the same value for all vertices. So it will never be interpolated, and will always be a constant value for every fragment inside a triangle. However WebGPU doesn't know what data I'm submitting so can't be sure, which is presumably why I get a warning.

There is some fairly impenetrable text around uniform control flow in the WGSL spec, so my question is basically, is this OK according to the spec? Or will this be made invalid in future and have to be done differently? If so how else should it be done? For example if it's made a separate attribute, it doesn't seem to change anything: it's a per-vertex value that we know is uniform but WebGPU doesn't know is uniform. I could try vertex pulling instead, but presumably that can't be statically analysed as uniform either as it will look like an arbitrary read from a buffer. Perhaps there could be some way of telling the WGSL compiler "yes I'm sure, this really is uniform"? What would be the consequence of breaking the uniformity?

**Comment by jrprice** [comment, marked as answer]
Hi Ashley, and thanks for the feedback! We've just recently implemented this feature in Chrome and are very interested in all feedback about the uniformity analysis.

These uniformity issues are currently flagged as warnings in Chrome, but will eventually become errors when we've ironed out any remaining issues and incorporated initial feedback. So you will indeed need to modify your shader code in order for it to continue compiling in future versions of Chrome (and other browsers, when they start implementing the same analysis).

If you're selecting between several different texture bindings, our recommended solution here would be to calculate derivatives explicitly and then use `textureSampleGrad`, which does not have the same uniformity constraints as `textureSample`. So, you could write something like this:

```
var texXy : vec2<f32> = input.fragUV.xy;
var texIndex : f32 = input.fragUV.z;
var c : vec4<f32>;

let dx = dpdx(texXy);
let dy = dpdx(texXy);

if (texIndex < 0.5)
{
	c = textureSampleGrad(texture0, sampler0, texXy, dx, dy)
}
else
{
	c = textureSampleGrad(texture1, sampler1, texXy, dx, dy)
}
```

This should be functionally equivalent, and performs the same number of accesses to the textures. An alternative solution is to just perform all of your `textureSample` calls unconditionally and then select between all of the outputs, but this may negatively impact performance if you have several different textures to choose from.

I hope this helps, and if you have any other feedback about the implementation or the impact of addressing these issues then please let us know.

**Reply by AshleyScirra** [reply]
Thanks for the workaround, I've added it in and it works fine and avoids the warning. I figure the problem is the `dpdx/y` calls must be in uniform control flow, and they're implicit in `textureSample`. The workaround moves them to uniform control flow, allowing just the texture sampling bit to happen in non-uniform control flow.

It would be nice if the error message could point out alternatives, e.g.

> warning: 'textureSample' must only be called from uniform control flow
> tip: consider using 'textureSampleGrad' which can be used with non-uniform control flow

I also don't know much about the low-level details of GPUs - out of curiosity why does this restriction exist for `dpdx/y`?

**Reply by Kangz** [reply]
GPUs run fragment shaders that use derivatives by groups of 2x2 called quads. This allows computing `dpdx/dpdy` easily by getting the value of the neighboring pixel in the quad and doing small computations. However if you are not in uniform control flow, then the neighboring pixel might not be in the same code path as the current invocation. It wouldn't call dpdx, or call it with a completely different variable, resulting in garbage values being used.

For example [`dFdx`](https://docs.gl/el3/dFdx) in GLSL call out:

> It is assumed that the expression p is continuous and therefore, expressions evaluated via non-uniform control flow may be undefined.

WGSL tries to help code more correct with the uniformity analysis with this restrictions (and avoid deadlocks when `workgroupBarrier` is called in non-uniform control flow).

**Reply by AshleyScirra** [reply]
Thanks for the explanation, I think I understand now. If fragments are rasterized in 2x2 quads, then all four fragments must be known to execute uniformly to avoid undefined behavior with derivatives. Makes sense to guarantee cross-platform consistency.
